vnext/JSI/Shared/ChakraObjectRef.cpp

```diff
@@ -0,0 +1,325 @@
+  if (m_state == State::Initialized) {
+    assert(m_ref);
+    CheckedJsRelease(m_ref);
+  } else {
+    assert(!m_ref);
+  }
+}
+
+void ChakraObjectRef::Initialize(JsRef ref) {
+  assert(!m_ref);
+
+  if (m_state != State::Uninitialized) {
+    throw facebook::jsi::JSINativeException(
+        "A ChakraObjectRef can only be initialzed once.");
+  }
+
+  if (!ref) {
+    throw facebook::jsi::JSINativeException(
+        "Cannot initialize a ChakraObjectRef with a null reference.");
+  }
+
+  CheckedJsAddRef(ref);
+  m_ref = ref;
+  m_state = State::Initialized;
+}
+
+void ChakraObjectRef::Invalidate() {
+  switch (m_state) {
+    case State::Uninitialized: {
+      assert(!m_ref);
+      throw facebook::jsi::JSINativeException(
+          "Cannot invalidate a ChakraObjectRef that has not been initialized.");
+      break;
+    }
+    case State::Initialized: {
+      assert(m_ref);
+      CheckedJsRelease(m_ref);
+      m_ref = nullptr;
+      m_state = State::Invalidated;
+      break;
+    }
+    case State::Invalidated: {
+      assert(!m_ref);
+      throw facebook::jsi::JSINativeException(
+          "Cannot invalidate a ChakraObjectRef that has already been "
+          "invalidated.");
+      break;
+    }
+    default: {
+      // Control flow should never reach here.
+      std::terminate();
+      break;
+    }
+  }
+}
+
+void ChakraObjectRef::swap(ChakraObjectRef &other) {
+  std::swap(m_ref, other.m_ref);
+  std::swap(m_state, other.m_state);
+}
+
+JsValueType GetValueType(const ChakraObjectRef &jsValue) {
+  JsValueType type;
+  ThrowUponChakraError(JsGetValueType(jsValue, &type), "JsGetValueType");
+  return type;
+}
+
+JsPropertyIdType GetPropertyIdType(const ChakraObjectRef &jsPropId) {
+  JsPropertyIdType type;
+  ThrowUponChakraError(
+      JsGetPropertyIdType(jsPropId, &type), "JsGetPropertyIdType");
+  return type;
+}
+
+std::wstring GetPropertyName(const ChakraObjectRef &id) {
+  if (GetPropertyIdType(id) != JsPropertyIdTypeString) {
+    throw facebook::jsi::JSINativeException(
+        "It is llegal to retrieve the name of a property symbol.");
+  }
+  const wchar_t *propertyName = nullptr;
+  ThrowUponChakraError(
+      JsGetPropertyNameFromId(id, &propertyName), "JsGetPropertyNameFromId");
+  return std::wstring{propertyName};
+}
+
+ChakraObjectRef GetPropertySymbol(const ChakraObjectRef &id) {
+  if (GetPropertyIdType(id) != JsPropertyIdTypeSymbol) {
+    throw facebook::jsi::JSINativeException(
+        "It is llegal to retrieve the symbol associated with a property name.");
+  }
+  JsValueRef symbol = nullptr;
+  ThrowUponChakraError(
+      JsGetSymbolFromPropertyId(id, &symbol), "JsGetSymbolFromPropertyId");
+  return ChakraObjectRef{symbol};
+}
+
+ChakraObjectRef GetPropertyId(const char *const utf8, size_t length) {
+  assert(utf8);
+
+  // We use a #ifdef here because we can avoid a UTF-8 to UTF-16 conversion
+  // using ChakraCore's JsCreatePropertyId API.
+#ifdef CHAKRACORE
+  JsPropertyIdRef id = nullptr;
```

Comment:
>JsPropertyIdRef [](start = 2, length = 15)

(nit) just curious, does JsPropertyIdRef constructor already set it to nullptr by any chance?

Reply:
JsPropertyIdRef is just a typedef for void*, so it follows the initialization rules of build-in types.

---
In reply to: [336174941](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336174941) [](ancestors = 336174941)